File: scripts/parse-foundry.js
#!/usr/bin/env node
/**
 * parse-foundry.js
 * Usage: node scripts/parse-foundry.js path/to/fvtt-Actor-*.json
 * Reads a Foundry VTT actor export and writes data/character.json
 */

const fs = require('fs');
const path = require('path');

const inputFile = process.argv[2];
if (!inputFile) {
  console.error('Usage: node scripts/parse-foundry.js <foundry-export.json>');
  process.exit(1);
}

const raw = JSON.parse(fs.readFileSync(inputFile, 'utf8'));
const sys = raw.system;

// --- Abilities ---
const abilityKeys = { str: 'FUE', dex: 'DES', con: 'CON', int: 'INT', wis: 'SAB', cha: 'CAR' };
const abilities = {};
for (const [key, label] of Object.entries(abilityKeys)) {
  const val = sys.abilities[key]?.value ?? 10;
  const mod = Math.floor((val - 10) / 2);
  abilities[key] = { value: val, mod, modStr: mod >= 0 ? `+${mod}` : `${mod}`, label, proficient: sys.abilities[key]?.proficient ?? 0 };
}

// --- Skills with proficiency ---
const skillMap = {
  acr: { name: 'Acrobatics', ability: 'dex' },
  ani: { name: 'Animal Handling', ability: 'wis' },
  arc: { name: 'Arcana', ability: 'int' },
  ath: { name: 'Athletics', ability: 'str' },
  dec: { name: 'Deception', ability: 'cha' },
  his: { name: 'History', ability: 'int' },
  ins: { name: 'Insight', ability: 'wis' },
  itm: { name: 'Intimidation', ability: 'cha' },
  inv: { name: 'Investigation', ability: 'int' },
  med: { name: 'Medicine', ability: 'wis' },
  nat: { name: 'Nature', ability: 'int' },
  prc: { name: 'Perception', ability: 'wis' },
  prf: { name: 'Performance', ability: 'cha' },
  per: { name: 'Persuasion', ability: 'cha' },
  rel: { name: 'Religion', ability: 'int' },
  slt: { name: 'Sleight of Hand', ability: 'dex' },
  ste: { name: 'Stealth', ability: 'dex' },
  sur: { name: 'Survival', ability: 'wis' },
};

// Get proficiency bonus from level
const classItem = raw.items.find(i => i.type === 'class');
const level = classItem?.system?.levels ?? 1;
const profBonus = Math.ceil(1 + level / 4);

const skills = {};
for (const [key, info] of Object.entries(skillMap)) {
  const sk = sys.skills[key];
  const abilMod = abilities[info.ability].mod;
  const profValue = sk?.value ?? 0; // 0=none, 1=prof, 2=expertise
  const bonus = abilMod + (profValue === 2 ? profBonus * 2 : profValue === 1 ? profBonus : 0);
  skills[key] = {
    name: info.name,
    ability: info.ability,
    proficient: profValue > 0,
    expertise: profValue === 2,
    bonus,
    bonusStr: bonus >= 0 ? `+${bonus}` : `${bonus}`,
  };
}

// --- HP ---
const hp = sys.attributes?.hp ?? {};

// --- Class & Subclass ---
const subclassItem = raw.items.find(i => i.type === 'subclass');
const className = classItem?.name ?? 'Unknown';
const subclassName = subclassItem?.name ?? '';

// --- Currency ---
const currency = sys.currency ?? {};

// --- Inspiration ---
const inspiration = sys.attributes?.inspiration ?? false;

// --- Exhaustion ---
const exhaustion = sys.attributes?.exhaustion ?? 0;

// --- Spells ---
const spells = raw.items
  .filter(i => i.type === 'spell')
  .map(i => ({
    name: i.name,
    level: i.system?.level ?? 0,
    prepMode: i.system?.preparation?.mode ?? 'prepared',
    school: i.system?.school ?? '',
  }))
  .sort((a, b) => a.level - b.level);

// --- Spell slots ---
const spellSlots = {};
for (let lvl = 1; lvl <= 9; lvl++) {
  const slot = sys.spells?.[`spell${lvl}`] ?? {};
  if (slot.max || slot.override || slot.value) {
    spellSlots[lvl] = { value: slot.value ?? 0, max: slot.override ?? slot.max ?? 0 };
  }
}

// --- Equipment ---
const equipmentTypes = ['weapon', 'equipment', 'consumable', 'tool', 'backpack', 'loot'];
const equipment = raw.items
  .filter(i => equipmentTypes.includes(i.type))
  .map(i => ({
    name: i.name,
    type: i.type,
    quantity: i.system?.quantity ?? 1,
    equipped: i.system?.equipped ?? false,
  }));

// --- Active effects ---
const effects = raw.items
  .filter(i => i.type === 'feat' || i.type === 'class' || i.type === 'subclass')
  .map(i => i.name);

const activeConditions = (raw.effects ?? [])
  .filter(e => !e.disabled)
  .map(e => e.name);

// --- Saving throws ---
const savingThrows = {};
for (const [key, ab] of Object.entries(abilities)) {
  const proficient = ab.proficient;
  const bonus = ab.mod + (proficient ? profBonus : 0);
  savingThrows[key] = { bonus, bonusStr: bonus >= 0 ? `+${bonus}` : `${bonus}`, proficient: proficient > 0 };
}

// --- Warnings ---
const warnings = [];
if (!hp.max) warnings.push({ type: 'error', msg: 'HP máximo no configurado en Foundry — ve a la hoja de personaje y configura manualmente.' });
if (Object.keys(spellSlots).length === 0) warnings.push({ type: 'error', msg: 'Spell slots no configurados en Foundry — revisa la pestaña de hechizos.' });
const hasDomainSpells = spells.some(s => ['Sleep', 'Faerie Fire'].includes(s.name));
if (!hasDomainSpells) warnings.push({ type: 'warn', msg: 'Los hechizos de dominio (Sleep, Faerie Fire) no están agregados en Foundry.' });
if (activeConditions.includes('Concentrating: Bless')) warnings.push({ type: 'info', msg: 'Concentrating: Bless está activo como efecto — desactívalo si hiciste descanso largo.' });

const armor = equipment.find(e => e.name.toLowerCase().includes('scale mail') || e.name.toLowerCase().includes('chain mail') || e.name.toLowerCase().includes('plate'));
if (armor?.name?.toLowerCase().includes('scale mail')) {
  warnings.push({ type: 'warn', msg: 'Tienes Scale Mail (CA 15 con escudo). Considera cambiar a Chain Mail (CA 18) — tienes proficiencia en armadura pesada por Twilight Domain.' });
}

// --- Output ---
const character = {
  name: raw.name,
  class: className,
  subclass: subclassName,
  level,
  background: raw.items.find(i => i.type === 'background')?.name ?? 'Outlander',
  race: raw.items.find(i => i.type === 'race')?.name ?? 'Dwarf (Hill)',
  deity: 'Selûne',
  proficiencyBonus: profBonus,
  abilities,
  savingThrows,
  skills,
  hp,
  spellSaveDC: 8 + profBonus + abilities.wis.mod,
  spellAttackBonus: profBonus + abilities.wis.mod,
  inspiration,
  exhaustion,
  currency,
  spells,
  spellSlots,
  equipment,
  activeConditions,
  warnings,
  lastUpdated: new Date().toISOString(),
};

const outputPath = path.join(__dirname, '../data/character.json');
fs.writeFileSync(outputPath, JSON.stringify(character, null, 2));
console.log(`✅ Character data written to ${outputPath}`);
console.log(`   Level ${level} ${className} (${subclassName})`);
console.log(`   HP: ${hp.value ?? '?'}/${hp.max ?? 'NOT SET'} | Inspiration: ${inspiration}`);
if (warnings.length > 0) {
  console.log(`\n⚠️  ${warnings.length} warning(s):`);
  warnings.forEach(w => console.log(`   [${w.type.toUpperCase()}] ${w.msg}`));
}

File: scripts/parse-foundry.py
#!/usr/bin/env python3
"""
parse-foundry.py
Reads a Foundry VTT actor export and writes data/character.json

Usage: python3 scripts/parse-foundry.py path/to/fvtt-Actor-adrik-*.json
"""

import json
import sys
import os
import math
from datetime import datetime

def fmt(n):
    return f"+{n}" if n >= 0 else str(n)

def mod(value):
    return math.floor((value - 10) / 2)

if len(sys.argv) < 2:
    print("Usage: python3 scripts/parse-foundry.py <foundry-export.json>")
    sys.exit(1)

input_file = sys.argv[1]
if not os.path.exists(input_file):
    print(f"❌ File not found: {input_file}")
    sys.exit(1)

with open(input_file, encoding="utf-8") as f:
    raw = json.load(f)

sys_data = raw.get("system", {})

# --- Abilities ---
ability_keys = {"str": "FUE", "dex": "DES", "con": "CON", "int": "INT", "wis": "SAB", "cha": "CAR"}
abilities = {}
for key, label in ability_keys.items():
    val = sys_data.get("abilities", {}).get(key, {}).get("value", 10)
    m = mod(val)
    prof = sys_data.get("abilities", {}).get(key, {}).get("proficient", 0)
    abilities[key] = {"value": val, "mod": m, "modStr": fmt(m), "label": label, "proficient": prof}

# --- Level & Proficiency ---
class_item = next((i for i in raw.get("items", []) if i.get("type") == "class"), None)
level = class_item.get("system", {}).get("levels", 1) if class_item else 1
prof_bonus = math.ceil(1 + level / 4)

# --- Skills ---
skill_map = {
    "acr": ("Acrobatics", "dex"), "ani": ("Animal Handling", "wis"),
    "arc": ("Arcana", "int"), "ath": ("Athletics", "str"),
    "dec": ("Deception", "cha"), "his": ("History", "int"),
    "ins": ("Insight", "wis"), "itm": ("Intimidation", "cha"),
    "inv": ("Investigation", "int"), "med": ("Medicine", "wis"),
    "nat": ("Nature", "int"), "prc": ("Perception", "wis"),
    "prf": ("Performance", "cha"), "per": ("Persuasion", "cha"),
    "rel": ("Religion", "int"), "slt": ("Sleight of Hand", "dex"),
    "ste": ("Stealth", "dex"), "sur": ("Survival", "wis"),
}
skills = {}
for key, (name, ability) in skill_map.items():
    sk = sys_data.get("skills", {}).get(key, {})
    a_mod = abilities[ability]["mod"]
    prof_val = sk.get("value", 0)
    bonus = a_mod + (prof_bonus * 2 if prof_val == 2 else prof_bonus if prof_val == 1 else 0)
    skills[key] = {
        "name": name, "ability": ability,
        "proficient": prof_val > 0, "expertise": prof_val == 2,
        "bonus": bonus, "bonusStr": fmt(bonus),
    }

# --- HP ---
hp = sys_data.get("attributes", {}).get("hp", {})

# --- Class & Subclass ---
subclass_item = next((i for i in raw.get("items", []) if i.get("type") == "subclass"), None)
class_name = class_item.get("name", "Unknown") if class_item else "Unknown"
subclass_name = subclass_item.get("name", "") if subclass_item else ""

# --- Background & Race ---
bg_item = next((i for i in raw.get("items", []) if i.get("type") == "background"), None)
race_item = next((i for i in raw.get("items", []) if i.get("type") == "race"), None)
background = bg_item.get("name", "Outlander") if bg_item else "Outlander"
race = race_item.get("name", "Dwarf (Hill)") if race_item else "Dwarf (Hill)"

# --- Currency, Inspiration, Exhaustion ---
currency = sys_data.get("currency", {})
inspiration = sys_data.get("attributes", {}).get("inspiration", False)
exhaustion = sys_data.get("attributes", {}).get("exhaustion", 0)

# --- Spells ---
spells = []
for i in raw.get("items", []):
    if i.get("type") != "spell":
        continue
    spells.append({
        "name": i.get("name", ""),
        "level": i.get("system", {}).get("level", 0),
        "prepMode": i.get("system", {}).get("preparation", {}).get("mode", "prepared"),
        "school": i.get("system", {}).get("school", ""),
    })
spells.sort(key=lambda x: x["level"])

# --- Spell Slots ---
spell_slots = {}
raw_spells = sys_data.get("spells", {})
for lvl in range(1, 10):
    slot = raw_spells.get(f"spell{lvl}", {})
    max_val = slot.get("override") or slot.get("max") or 0
    cur_val = slot.get("value") or 0
    if max_val or cur_val:
        spell_slots[lvl] = {"value": cur_val, "max": max_val}

# --- Equipment ---
equipment_types = {"weapon", "equipment", "consumable", "tool", "backpack", "loot"}
equipment = []
for i in raw.get("items", []):
    if i.get("type") not in equipment_types:
        continue
    equipment.append({
        "name": i.get("name", ""),
        "type": i.get("type", ""),
        "quantity": i.get("system", {}).get("quantity", 1),
        "equipped": i.get("system", {}).get("equipped", False),
    })

# --- Saving Throws ---
saving_throws = {}
for key, ab in abilities.items():
    prof = ab["proficient"]
    bonus = ab["mod"] + (prof_bonus if prof else 0)
    saving_throws[key] = {"bonus": bonus, "bonusStr": fmt(bonus), "proficient": prof > 0}

# --- Active Conditions ---
active_conditions = [e.get("name", "") for e in raw.get("effects", []) if not e.get("disabled", True)]

# --- Derived Stats ---
spell_save_dc = 8 + prof_bonus + abilities["wis"]["mod"]
spell_attack_bonus = prof_bonus + abilities["wis"]["mod"]

# --- Warnings ---
warnings = []
if not hp.get("max"):
    warnings.append({"type": "error", "msg": "HP máximo no configurado en Foundry — configúralo manualmente en la hoja de personaje."})
if not spell_slots:
    warnings.append({"type": "error", "msg": "Spell slots no configurados en Foundry — revisa la pestaña de hechizos."})
domain_spells = {"Sleep", "Faerie Fire"}
has_domain = any(s["name"] in domain_spells for s in spells)
if not has_domain:
    warnings.append({"type": "warn", "msg": "Los hechizos de dominio (Sleep, Faerie Fire) no están agregados en Foundry."})
if any("scale mail" in e["name"].lower() for e in equipment):
    warnings.append({"type": "warn", "msg": "Tienes Scale Mail (CA 15). Considera cambiar a Chain Mail (CA 18) — tienes proficiencia en armadura pesada por Twilight Domain."})

# --- Build output ---
character = {
    "name": raw.get("name", ""),
    "class": class_name,
    "subclass": subclass_name,
    "level": level,
    "background": background,
    "race": race,
    "deity": "Selûne",
    "proficiencyBonus": prof_bonus,
    "abilities": abilities,
    "savingThrows": saving_throws,
    "skills": skills,
    "hp": hp,
    "spellSaveDC": spell_save_dc,
    "spellAttackBonus": spell_attack_bonus,
    "inspiration": inspiration,
    "exhaustion": exhaustion,
    "currency": currency,
    "spells": spells,
    "spellSlots": {str(k): v for k, v in spell_slots.items()},
    "equipment": equipment,
    "activeConditions": active_conditions,
    "warnings": warnings,
    "lastUpdated": datetime.now().isoformat(),
}

output_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data", "character.json")
os.makedirs(os.path.dirname(output_path), exist_ok=True)
with open(output_path, "w", encoding="utf-8") as f:
    json.dump(character, f, indent=2, ensure_ascii=False)

print(f"✅ Character data written to {output_path}")
print(f"   Level {level} {class_name} ({subclass_name})")
print(f"   HP: {hp.get('value','?')}/{hp.get('max','NOT SET')} | Inspiration: {inspiration}")
if warnings:
    print(f"\n⚠️  {len(warnings)} warning(s):")
    for w in warnings:
        print(f"   [{w['type'].upper()}] {w['msg']}")
